blocks/algorithms.py

```diff
@@ -0,0 +1,227 @@
+"""Training algorithms."""
+from abc import ABCMeta, abstractmethod
+from collections import OrderedDict
+
+import theano
+from theano import tensor
+
+from blocks.utils import update_instance
+from blocks.graph import ComputationGraph
+
+
+class TrainingAlgorithm(object):
+    """Base class for training algorithms.
+
+    A training algorithm object has a simple life-cycle.
+    First it is initialized by calling its :meth:`initialize` method.
+    At this stage, for instance, Theano functions can be compiled.
+    After that the :meth:`process_batch` method is repeatedly
+    called with a batch of training data as a parameter.
+
+    """
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def initialize(self):
+        """Initialize the training algorithm."""
+        pass
+
+    @abstractmethod
+    def process_batch(self, batch):
+        """Process a batch of training data.
+
+        Attributes
+        ----------
+        batch : dict
+            A dictionary of (source name, data) pairs.
+
+        """
+        pass
+
+
+class TheanoTrainingAlgorithm(TrainingAlgorithm):
+    """A training algorithm based on Theano functions.
+
+    Parameters
+    ----------
+    cost : Theano variable
+        The objective to be minimized.
+    params : list of Theano shared variables
+        The parameters to be tuned.
+
+    Attributes
+    ----------
+    updates : list of (shared variable, Theano expression) tuples
+        Updates to be done for every batch. It is required that the
+        updates are done using the old values of optimized parameters.
+    cost : Theano variable
+        The objective to be minimized.
+    params : list of Theano shared variables
+        The parameters to be tuned.
+
+    Notes
+    -----
+        Changing `updates` attribute or calling `add_updates` after
+        the `initialize` method is called will have no effect.
+
+    .. todo::
+
+       Ideally the parameters should be extracted from the `cost`
+       computation graph. But in order for this to work our graph
+       traversal routine should recurse into ScanOp's, which is currently
+       not the case.
```

Comment:
I would actually prefer it if the main loop is in charge of collecting the
parameters in all cases. I don't like the idea of having "soft
dependencies" everywhere; they tend to obfuscate functionality, and turn
into hard dependencies unnoticed.

I understand the desire for a minimal main loop, but I think we should
achieve this by pushing functionality into training extensions, not into
other parts of the library. Collecting parameters should never be the job
of a training algorithm.
On Jan 15, 2015 8:14 AM, "Dmitry Bogdanov" <notifications@github.com> wrote:

> In blocks/algorithms.py
> <https://github.com/bartvm/blocks/pull/92#discussion-diff-23008211>:
>
> > +    cost : Theano variable
> > +        The objective to be minimized.
> > +    params : list of Theano shared variables
> > +        The parameters to be tuned.
> > +
> > +    Notes
> > +    -----
> > +        Changing `updates` attribute or calling `add_updates` after
> > +        the `initialize` method is called will have no effect.
> > +
> > +    .. todo::
> > +
> > +       Ideally the parameters should be extracted from the `cost`
> > +       computation graph. But in order for this to work our graph
> > +       traversal routine should recurse into ScanOp's, which is currently
> > +       not the case.
>
> Agreed. If params are None, they can be retrieved from the computation
> graph, and this will be done in most cases.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bartvm/blocks/pull/92/files#r23008211>.
>